### DISCUSSION TITLE
Getting notified when algorithm is terminated

### DISCUSSION THREAD

**Discussion by ThomasEwaldETH** [anyoptimization/pymoo#398] [Q&A] [answered]
Hello everyone!
First, thanks for maintaining this great package! It is very convenient and helpful for me!

To my optimization problem:
During my optimization, I use the Callback class to get notified every time a generation is evaluated. I then take these results and put them into a plot, so that my algorithm does some sort of 'runtime plotting' of the status of the optimization, and I can see how the Pareto front forms. This works very smoothly and is exactly what I need. In addition, I show the current iteration number as the plot title. Here is an example:

    classPlotStandalone(Callback):
        def __init__(self) -> None:
            super().__init__()
            self.fig = plt.figure(1)
            self.axes = self.fig.add_subplot()

        def notify(self, algorithm: NSGA2) -> None:
            population = algorithm.pop
            iteration = algorithm.n_iter
            values = population.get("F")
            x = values[:, 0]
            y = values[:, 1]
            self.plot(x, y, iteration)

        def plot(self, x: "npt.NDArray", y: "npt.NDArray", it: int) -> None:
            ... # plots in self.axes

To my question:
I was wondering whether I can check if the algorithm has terminated? After finishing, I want to plot some additional information. My actual code runs the optimization with the minimize() method. When finished, I access the results object and do this “manually”. However, I would prefer a solution where my notify method realizes the termination and can call another function, that does the post-processing. The reason is, that the post-processing is again changing stuff in the plot (plotting my favorite solution) and I want to have that inside the Callback class, like this:

    classPlotStandalone(Callback):
        def __init__(self) -> None:
            ...

        def notify(self, algorithm: NSGA2) -> None:
            if terminated:
                ...
                self.has_terminated(...)
            else:
                ...
                self.plot(x, y, iteration)

        def plot(self, ...) -> None:
            ...

        def has_terminated(self, ...) -> None:
            ...

So, is that possible? Is there something like a flag in the Termination object, that indicates "final generation reached" or something?

Best regards and have a nice day,
Thomas

**Comment by cyrilpic** [comment, marked as answer]
The callback is called after the termination object has been updated. So, in your notify method, you can check if `algorithm.termination.has_terminated()` is True.

**Reply by ThomasEwaldETH** [reply]
That's great, thanks a lot (tested it in my code, and it works exactly like I had it in mind)!
Best regards,
Thomas

**Comment by blankjul** [comment]
I have nothing to add. I would have suggested exactly the same.

@cyrilpic Thank you very much for helping out here :)
